cranelift/frontend/src/ssa.rs

```diff
@@ -423,46 +427,35 @@ impl SSABuilder {
         // and creation of new blocks, however such new blocks are sealed on
         // the fly, so we don't need to account for them here.
         for block in self.ssa_blocks.keys() {
-            if !self.is_sealed(block) {
-                self.seal_one_block(block, func);
-            }
+            self.seal_one_block(block, func);
         }
-        mem::replace(&mut self.side_effects, SideEffects::new())
+        mem::take(&mut self.side_effects)
     }
 
-    /// Helper function for `seal_block` and
-    /// `seal_all_blocks`.
+    /// Helper function for `seal_block` and `seal_all_blocks`.
     fn seal_one_block(&mut self, block: Block, func: &mut Function) {
-        let block_data = &mut self.ssa_blocks[block];
-        debug_assert!(
-            !block_data.sealed,
-            "Attempting to seal {} which is already sealed.",
-            block
-        );
-
-        // Extract the undef_variables data from the block so that we
-        // can iterate over it without borrowing the whole builder.
-        let undef_vars = mem::replace(&mut block_data.undef_variables, Vec::new());
-
         // For each undef var we look up values in the predecessors and create a block parameter
         // only if necessary.
-        for (var, val) in undef_vars {
-            let ty = func.dfg.value_type(val);
-            self.predecessors_lookup(func, val, var, ty, block);
+        for (var, val) in self.mark_block_sealed(block) {
+            debug_assert!(self.calls.is_empty());
+            debug_assert!(self.results.is_empty());
+            // self.side_effects may be non-empty here so that callers can
+            // accumulate side effects over multiple calls.
+            self.begin_predecessors_lookup(val, block);
+            self.run_state_machine(func, var, func.dfg.value_type(val));
         }
-        self.mark_block_sealed(block);
     }
 
-    /// Set the `sealed` flag for `block`.
-    fn mark_block_sealed(&mut self, block: Block) {
-        let block_data = &mut self.ssa_blocks[block];
-        debug_assert!(!block_data.sealed);
-        debug_assert!(block_data.undef_variables.is_empty());
-        block_data.sealed = true;
-
+    /// Set the `sealed` flag for `block`. Returns any variables that still need definitions.
+    fn mark_block_sealed(&mut self, block: Block) -> Vec<(Variable, Value)> {
         // We could call data.predecessors.shrink_to_fit() here, if
         // important, because no further predecessors will be added
         // to this block.
+
+        match mem::replace(&mut self.ssa_blocks[block].sealed, Sealed::Yes) {
+            Sealed::No { undef_variables } => undef_variables,
+            Sealed::Yes => Vec::new(),
```

Comment:
I think this results in a change in behavior with respect to before: earlier we had an assert that ensured a block could only be sealed once (line 437 in the original); now we silently accept `Yes` here. That is probably OK -- the effect is nothing if already sealed -- but I just want to make sure we've thought about this and that it doesn't contradict docs elsewhere?

Reply:
In terms of public API, the behavior is unchanged. I moved the assert to `seal_block`; and `seal_all_blocks` was only calling `seal_one_block` if the block was not already sealed.

Making the internal helpers just no-op for sealed blocks meant that I could remove a conditional from `seal_all_blocks` and check only once in `mark_block_sealed` instead.

As far as private callers go, `append_jump_argument` calls `mark_block_sealed` but only on a block it just created, so it definitely isn't sealed.